File: Tests/AICallAssistantTests/ContextLibraryStoreTests.swift
import Foundation
import XCTest
@testable import AICallAssistant

final class ContextLibraryStoreTests: XCTestCase {
    func testLoadReturnsEmptyArrayWhenContextsFileDoesNotExist() throws {
        let rootURL = makeTemporaryRoot()
        defer { try? FileManager.default.removeItem(at: rootURL) }

        let store = ContextLibraryStore(rootURL: rootURL)

        XCTAssertEqual(try store.load(), [])
    }

    func testSaveAndLoadRoundTripsCompleteContextLibraryInOrder() throws {
        let rootURL = makeTemporaryRoot()
        defer { try? FileManager.default.removeItem(at: rootURL) }
        let store = ContextLibraryStore(rootURL: rootURL)
        let contexts = [
            CallContext(
                id: UUID(uuidString: "10000000-0000-0000-0000-000000000001")!,
                title: "Резюме 👩🏽‍💻",
                body: "Первая строка\nВторая строка — опыт в B2B SaaS.",
                isSelected: false,
                attachments: [
                    ContextFileAttachment(
                        id: UUID(uuidString: "20000000-0000-0000-0000-000000000001")!,
                        fileName: "CV Андрея 📄.pdf",
                        mediaType: "application/pdf",
                        byteCount: 12_345,
                        contentSHA256: String(repeating: "a", count: 64),
                        extractedText: "Точный текст файла\nс Unicode: 你好, مرحبا, 🙂"
                    )
                ]
            ),
            CallContext(
                id: UUID(uuidString: "10000000-0000-0000-0000-000000000002")!,
                title: "Вакансия",
                body: "Head of Product",
                isSelected: true,
                attachments: []
            )
        ]

        try store.save(contexts)

        let contextsURL = rootURL.appendingPathComponent("contexts.json")
        XCTAssertTrue(FileManager.default.fileExists(atPath: contextsURL.path))
        XCTAssertEqual(try permissions(of: rootURL), 0o700)
        XCTAssertEqual(try permissions(of: contextsURL), 0o600)
        XCTAssertEqual(try store.load(), contexts)
    }

    func testSavingEmptyLibraryOverwritesPreviouslySavedContexts() throws {
        let rootURL = makeTemporaryRoot()
        defer { try? FileManager.default.removeItem(at: rootURL) }
        let store = ContextLibraryStore(rootURL: rootURL)
        let context = CallContext(
            title: "Контекст, который удалили",
            body: "Не должен появиться после перезапуска",
            isSelected: true
        )

        try store.save([context])
        try store.save([])

        XCTAssertEqual(try store.load(), [])
    }

    func testLoadThrowsForUnsupportedSchemaVersion() throws {
        let rootURL = makeTemporaryRoot()
        defer { try? FileManager.default.removeItem(at: rootURL) }
        try FileManager.default.createDirectory(
            at: rootURL,
            withIntermediateDirectories: true
        )
        let contextsURL = rootURL.appendingPathComponent("contexts.json")
        try Data(
            """
            {
              "schemaVersion": 999,
              "contexts": []
            }
            """.utf8
        ).write(to: contextsURL, options: .atomic)
        let store = ContextLibraryStore(rootURL: rootURL)

        XCTAssertThrowsError(try store.load())
    }

    func testRecoverySavePreservesUnreadableLibraryBeforeReplacement() throws {
        let rootURL = makeTemporaryRoot()
        defer { try? FileManager.default.removeItem(at: rootURL) }
        try FileManager.default.createDirectory(
            at: rootURL,
            withIntermediateDirectories: true
        )
        let contextsURL = rootURL.appendingPathComponent(ContextLibraryStore.filename)
        let unreadableData = Data("future-or-corrupt-library".utf8)
        try unreadableData.write(to: contextsURL, options: .atomic)
        let store = ContextLibraryStore(rootURL: rootURL)
        let replacement = [
            CallContext(title: "Recovered edit", body: "New library", isSelected: true)
        ]

        try store.save(replacement, preservingExistingAsRecovery: true)

        let recoveryURLs = try FileManager.default.contentsOfDirectory(
            at: rootURL,
            includingPropertiesForKeys: nil
        ).filter {
            $0.lastPathComponent.hasPrefix(ContextLibraryStore.recoveryFilenamePrefix)
        }
        XCTAssertEqual(recoveryURLs.count, 1)
        XCTAssertEqual(try Data(contentsOf: try XCTUnwrap(recoveryURLs.first)), unreadableData)
        XCTAssertEqual(try permissions(of: try XCTUnwrap(recoveryURLs.first)), 0o600)
        XCTAssertEqual(try store.load(), replacement)
    }

    private func makeTemporaryRoot() -> URL {
        FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString, isDirectory: true)
            .appendingPathComponent("AI Call Assistant", isDirectory: true)
    }

    private func permissions(of url: URL) throws -> Int {
        let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
        return try XCTUnwrap(attributes[.posixPermissions] as? NSNumber).intValue & 0o777
    }
}

File: windows/tests/AICallAssistant.Core.Tests/ContextLibraryStoreTests.cs
using AICallAssistant.Core.Models;
using AICallAssistant.Core.Persistence;

namespace AICallAssistant.Core.Tests;

public sealed class ContextLibraryStoreTests
{
    [Fact]
    public async Task MissingLibraryLoadsAsEmpty()
    {
        using var temporaryDirectory = new TestDirectory();
        var store = new ContextLibraryStore(temporaryDirectory.RootPath);

        var contexts = await store.LoadAsync();

        Assert.Empty(contexts);
    }

    [Fact]
    public async Task CompleteLibraryRoundTripsInOrder()
    {
        using var temporaryDirectory = new TestDirectory();
        var store = new ContextLibraryStore(temporaryDirectory.RootPath);
        var firstId = Guid.Parse("10000000-0000-0000-0000-000000000001");
        var attachmentId = Guid.Parse("20000000-0000-0000-0000-000000000001");
        var contexts = new[]
        {
            new CallContext
            {
                Id = firstId,
                Title = "Резюме 👩🏽‍💻",
                Body = "Первая строка\nВторая строка",
                IsSelected = true,
                Attachments =
                [
                    new ContextFileAttachment
                    {
                        Id = attachmentId,
                        FileName = "CV Андрея 📄.pdf",
                        MediaType = "application/pdf",
                        ByteCount = 12_345,
                        ContentSha256 = new string('a', 64),
                        ExtractedText = "Unicode: 你好, مرحبا, 🙂"
                    }
                ]
            },
            new CallContext
            {
                Title = "Вакансия",
                Body = "Head of Product"
            }
        };

        await store.SaveAsync(contexts);
        var reloaded = await store.LoadAsync();

        Assert.Equal(2, reloaded.Count);
        Assert.Equal(firstId, reloaded[0].Id);
        Assert.Equal("Резюме 👩🏽‍💻", reloaded[0].Title);
        Assert.True(reloaded[0].IsSelected);
        var attachment = Assert.Single(reloaded[0].Attachments);
        Assert.Equal(attachmentId, attachment.Id);
        Assert.Equal("Unicode: 你好, مرحبا, 🙂", attachment.ExtractedText);
        Assert.Equal("Вакансия", reloaded[1].Title);
    }

    [Fact]
    public async Task CorruptLibraryIsMovedToRecoveryBeforeStartingEmpty()
    {
        using var temporaryDirectory = new TestDirectory();
        var store = new ContextLibraryStore(temporaryDirectory.RootPath);
        var corruptBytes = "future-or-corrupt-library"u8.ToArray();
        await File.WriteAllBytesAsync(store.FilePath, corruptBytes);

        var contexts = await store.LoadAsync();

        Assert.Empty(contexts);
        Assert.False(File.Exists(store.FilePath));
        var recoveryPath = Assert.Single(
            Directory.EnumerateFiles(temporaryDirectory.RootPath),
            path => Path.GetFileName(path).StartsWith(
                    ContextLibraryStore.RecoveryFilePrefix,
                    StringComparison.Ordinal));
        Assert.Equal(corruptBytes, await File.ReadAllBytesAsync(recoveryPath));

        await store.SaveAsync([new CallContext { Title = "Recovered edit" }]);
        Assert.Equal("Recovered edit", Assert.Single(await store.LoadAsync()).Title);
    }
}
